### PR TITLE
fixed issue #559

### LINKED ISSUE #559: CLI does not close when pressing X to exit the dev CLI
### Description:

When you're in the dev CLI, you're prompted to press X to close the dev server.
That does kill the containers, but it doesn't fully detach from the process.

Image (credit: @thesujai)
image

### Steps to reproduce:

1. Ensure dev server is running via ./dev.sh
2. Wait for everything to start fully.
3. Once ready, press `x`.

### Expected behavior:

All containers are killed ✅ 
Process detaches cleanly from the terminal ❌ 

### Actual behavior:

As shared in the screenshot above, process does not detach cleanly from the terminal.

### PR DESCRIPTION
**Pull request by shamun-khatri**
## Linked Issue(s)

Closes #559 

## Proposed changes (including videos or screenshots)

I have used process.exit() in one place (useEffect hook where it checks if  AppState === TERMINATED) instead of exit() that is not working. I have tried to solve this issue by debugging and trying other methods but noting works but this.

Here is a screenshot of working terminal detach.

image

## Further comments

I hope this will get merged. But if you think I could do better then tell me I will change things along.

### CODE REVIEW

**Review by jayantbh — CHANGES_REQUESTED** [review]
I have a feeling that this doesn't always work.
For example, if you let the ./dev.sh script run till completion (or long enough), and then press X, it doesn't quit successfully.

I noticed you replaced `ink`'s `exit` call with `process.exit`, which is fine, but the dependencies remain unchanged.

I think I'd have kept both the `process.exit` and `exit` calls in there.

Additionally...
There's a lot of whitespace changes which I'm not sure are intentional.

image

**Comment by shamun-khatri** [comment]
Yes, whitespaces are not intentional I have tried many other things that's why they are there.

But I got confused about what will not work.  You said in a long-running process that might now work why?  

Also, what does it mean that dependencies remain unchanged?

**Comment by shamun-khatri** [comment]
One thing I forgot about I have tried keeping both `exit()` and `process.exit()` and keeping both also works.

And I have also tried keeping the app(process) running for 5 minutes and then tried X with only `process.exit()` and it still works fine.

**Comment by jayantbh** [comment]
> But I got confused about what will not work.
Honestly, same. 😆 

I wasn't able to reliably reproduce that issue in the past.

But yeah, I think call `exit` and then `process.exit`. Call both. `exit` from `ink` allows it to do any needed cleanup before actual exit.

And do remove any unintended code/whitespace changes. I think that'll be it for this PR.

**Comment by thesujai** [comment]
I hovered around inks codebase and found the reason exit() isn't working is some processes are still active. So the exit() by useApp isn't detaching the terminal. 
The processref.current.kill() call in the handleExit is not working as intended. 

So this solution introduced here might be a hacky way to do things, but it doesn't break anything. Just we can do a check that after calling exit(), if the process is still not killed, then call the process.exit(0) (to keep in mind this isn't forced as code 0 is to safely kill the process). This will ensure that no process is running after the process is unmounted. Else failure to do so might result in terminal being unmounted, but some services still running in the bg.

**Comment by jayantbh** [comment]
Interesting. Yeah I can see how that happens.
How do we check if a process hasn't safely and completely exited?

**Comment by thesujai** [comment]
After exit() call in useEffect(), can do:
 ```
if(processRef.current) {
process.exit(0);
}
```

**Comment by thesujai** [comment]
> But why some process is still running there after running Ikn's exit()

Try running git blame on this file and see what introduced this bug(Maybe it worked before, not sure)

> are you raising a PR for this

No, and feel free to update this branch. Good job!

File: cli/source/app.tsx [hunk 1]
```diff
@@ -132,6 +132,9 @@ const CliUi = () => {
   useEffect(() => {
     if (appState !== AppStates.TERMINATED) return;
     exit();
+    if(processRef.current) {
+      process.exit(0);
+      }
```
Comment (jayantbh) [inline cli/source/app.tsx:137]: Why the `processRef` check here?
Also, this doesn't look formatted right.

**Review by jayantbh — CHANGES_REQUESTED** [review]

**Comment by jayantbh** [comment]
I understand the intent, but given that you didn't really do anything with processRef itself, and that `exit` is being called anyway, I think it's okay to just `kill` it.

A check for `processRef` would fit better if we were doing something with it directly or indirectly.

Right now if `processRef` happens to not be set, then the UI is potentially just not going to exit properly, which isn't desired behaviour either.

**Comment by shamun-khatri** [comment]
Ok, I will just kill that and remove unnecessary checks.I have joined the community on slack thanks for sharing.

**Comment by shamun-khatri** [comment]
I made changes. Please have a look.

**Comment by jayantbh** [comment]
image
@shamun-khatri, please use better commit names to better reflect the changes within.

I wrote about this a while ago.
https://dev.to/middleware/not-heres-how-to-write-actually-good-commit-messages-hint-its-not-just-adding-commit-lint-j2i

**Comment by shamun-khatri** [comment]
@jayantbh I have renamed the commit. I tried my best to write that but if you thing I am still missing anything then do let me know.

**Comment by jayantbh** [comment]
@shamun-khatri does it count now? I assigned the PR to you, which I think you can do yourself too.

**Review by jayantbh — APPROVED** [review]

**Comment by shamun-khatri** [comment]
@jayantbh Thanks, finally it is merged. will look forward to contributing more to middleware.

**Pull request merged**